snuba/clickhouse/astquery.py

```diff
@@ -0,0 +1,150 @@
+        self.__prewhere = query.get_prewhere_ast()
+
+        self.__settings = settings
+        self.__formatted_query: Optional[str] = None
+
+    def format_sql(self) -> str:
+        if self.__formatted_query:
+            return self.__formatted_query
+
+        parsing_context = ParsingContext()
+        formatter = ClickhouseExpressionFormatter(parsing_context)
+
+        selected_cols = [e.accept(formatter) for e in self.__selected_columns]
+        formatted_columns = ", ".join(selected_cols)
+        select_clause = f"SELECT {formatted_columns}"
+
+        # TODO: The visitor approach will be used for the FROM clause as well.
+        from_clause = f"FROM {self.__data_source.format_from()}"
+
+        if self.__final:
+            from_clause = f"{from_clause} FINAL"
+
+        # TODO: Sampling rate will become one step of Clickhouse query processing
+        if not self.__data_source.supports_sample():
+            sample_rate = None
+        else:
+            if self.__sample:
+                sample_rate = self.__sample
+            elif self.__settings.get_turbo():
+                sample_rate = settings.TURBO_SAMPLE_RATE
+            else:
+                sample_rate = None
+        if sample_rate:
+            from_clause = f"{from_clause} SAMPLE {sample_rate}"
+
+        array_join_clause = ""
+        if self.__arrayjoin:
+            formatted_array_join = self.__arrayjoin.accept(formatter)
+            array_join_clause = f"ARRAY JOIN {formatted_array_join}"
+
+        prewhere_clause = ""
+        if self.__prewhere:
+            formatted_prewhere = self.__prewhere.accept(formatter)
+            prewhere_clause = f"PREWHERE {formatted_prewhere}"
+
+        where_clause = ""
+        if self.__condition:
+            formatted_condition = self.__condition.accept(formatter)
+            where_clause = f"WHERE {formatted_condition}"
+
+        group_clause = ""
+        if self.__groupby:
+            # reformat to use aliases generated during the select clause formatting.
+            groupby_expressions = [e.accept(formatter) for e in self.__groupby]
+            formatted_groupby = ", ".join(groupby_expressions)
+            group_clause = f"GROUP BY ({formatted_groupby})"
+            if self.__hastotals:
+                group_clause = f"{group_clause} WITH TOTALS"
+
+        having_clause = ""
+        if self.__having:
+            formatted_having = self.__having.accept(formatter)
+            having_clause = f"HAVING {formatted_having}"
+
+        order_clause = ""
+        if self.__orderby:
+            orderby = [
+                f"{e.node.accept(formatter)} {e.direction.value}"
+                for e in self.__orderby
+            ]
+            formatted_orderby = ", ".join(orderby)
+            order_clause = f"ORDER BY {formatted_orderby}"
+
+        limitby_clause = ""
+        if self.__limitby is not None:
+            limitby_clause = "LIMIT {} BY {}".format(*self.__limitby)
```

Comment:
Should it also be f string?

Reply:
Generally agree with this, it'd be nice if `Limitby` was a namedtuple or dataclass and this was referenced by attribute names instead of unpacking.

Reply:
Unfortunately this comes as a tuple because it is a tuple in the Snuba syntax. We need to change the structure in the parsing process to have something different here.